### ISSUE TITLE
[Fastboot support] Integrate building & serving fastboot in ember-cli

### ISSUE THREAD

**Issue by kratiahuja** [ember-cli/ember-cli#6350]
This is a master issue tracking the work to integrate building fastboot in ember-cli. We can break this issue into smaller ones if needed. Note: This is a backward compatible proposal in `ember-cli` but incompatible with current fastboot ergonomics. 

**TL;DR:** After this issue is fixed it will allow:
1. Fastboot assets to be built using `ember build` and be performant.
2. Allow a different server (in this case fastboot server) to serve files rather than `ember-cli` serve middleware using `ember serve`.
3. Kill `ember fastboot` command itself.

I have a prototype and some of the changes working (building `fastboot/app` and `fastboot/addon`) with `ember-cli`.

**Thanks to @stefanpenner  for the idea!**
## Motivation

[Fastboot](https://ember-fastboot.com) allows ember apps to be rendered on the server side. It renders the initial content of your app allowing your app to be used in SEO usecases. On a very high level, fastboot loads your assets in a Node environment and creates the Ember App once the assets are loaded. It does not do any automatic routing but creates Ember application instance whenever there is a server side rendering request and runs your route to render the initial render in Node.

In order for your app/addon to be able to run in a Node environment, it needs to be compatible in such environment. Compatibility includes not doing DOM/window access in Node, making sure you are able to make API requests in Node etc. Therefore an app/addon may need to provide fastboot specific initializers, instance-initializers, services etc that are only applicable in the Node environment and not in the browser. Including them in the browser has performance implication since you will send down more bytes to the client.

Currently fastboot builds the app twice and generates a different set of assets for fastboot and browser environments via `ember fastboot` command. Most of the app between the two sets of assets is same except where we want fastboot to override the browser behavior or need fastboot specific handling. Currently `ember fastboot` filters these and generates the two sets of assets. Building different sets of assets (app and vendor files) for different environments cannot guarantee the app works correctly in both environments always. Moreover, `ember-fastboot --serve-assets` command also spins up a node server and serves your assets. This is very similar to what `ember serve` does today. However, we end up forking and creating two parallel worlds with problems such as serving stale assets, creating race conditions during serving and building etc. We want to be able converge the two worlds such that fastboot should be able to specify the last middleware that will be used to serve the assets in fastboot. This helps us not create fastboot commands and keeps the eco-system unified.

We want to build these fastboot specific code as an additive asset to the browser that loads in the Node VM in addtion to the the other assets. If it overrides any browser behavior, having the same module id will override the definition in the loader. In order to build fastboot assets in a performant way, we would like to integrate building fastboot into ember-cli. This issue specifically tracks integrating building fastboot in ember-cli. When we need to build more such artifacts, we can extend this to be a generic target solution.
## Solution

In order to build the fastboot assets which are an addition to the app and vendor files, we want to define a separate directory structure. This will allow fastboot specific app/addon code to live in its own namespace and does not touch the app and vendor files.

For example an addon containing fastboot specific functionality will have a directory structure like:

```
my-addon/
|-- addon/
│   |-- utils/
│   │   |-- foo.js
│-- app/
│   │-- instance-initializers/
│   │   |-- error-handler.js
│-- fastboot/
│   │-- app/
│   │   |-- instance-initializers/
|   |   |   |-- error-handler.js
|   |   |   |-- ajax.js
|   |-- addon/
|   |   |-- utils/
|   |   |   |-- foo.js
|   |   |   |-- fastboot-utils.js
```

Note: The same applies within an app as well, just that it wouldn't have `fastboot/addon` folder.

This allows us to keep fastboot specific functionality in its own namespace and allows us to build fastboot artifact by walking only the subset of the tree rather than walking the whole app and doing filtering (as done right now).

After an `ember build`, the generated assets will be:
1. vendor.js
2. app.js
3. app-fastboot.js

In the Node VM, we will first load the assets and eval them in the above order. Therefore, if you wanted to override a functionality in fastboot, you can keep the name of the file same in fastboot namespace and will override it when `app-fastboot.js` gets loaded and evaluated. This also guarantees that we do not regress existing ember builds that do not contain the fastboot namespace.
## Tasks

This issue tracks the list of changes and public APIs we need to expose for supporting the above strategy in `ember-cli`:
- [ ] Support building fastboot/app from app and addons and generate `<app>-fastboot.js`
- [ ] Make sure `broccoli-asset-rev` is able to fingerprint these additional assets
- [ ] Add `fastboot` as a type of tree in `preProcessTree` and `postprocessTree`
- [ ] Expose `treeForFastbootApp` for fastboot addons to add/remove items from fastboot specific trees. This will allow us to add HTMLbars munging based on Ember versions.
- [ ] Make sure JSHint is enabled for `fastboot/`
- [ ] Expose an API from `ember-cli` to allow an addon to add a middleware to serve assets (See issue [here](https://github.com/ember-fastboot/ember-cli-fastboot/issues/274)). I still need to explore this a bit more.
- [ ] Make sure live reload works with fastboot changes.
- [ ] Update documentation of ember-cli with the additonal public hooks and building of fastboot assets.
- [ ] Update documentation of `ember-fastboot`

## Future

We may want to create different types of additional assets in future that are addition to the existing assets. We may want to make it generic such that we could say `ember serve --target fastboot`.
## Working on this issue?

I am going to send out a `[WIP]` PR fixing the tasks in the above list. Please feel free to pick any and add it as a comment. 

cc: @tomdale @stefanpenner  @danmcclain  @arjan @masonwan @ryanone

**Comment by arjansingh** [comment]
Generally I like this. I have no major conceptual issues with it on first read. 

Points of feedback:

1. The new directory structure would be a big breaking change for Ember Apps. Does that mean we're at `2.0.0` already?
2.  "Explicitly disallow styles in `fastboot/`"... I'm not sure I understand the issue. Could you elaborate?
3.  Our testing strategy for FastBoot has been a little ad-hoc. I do think `ember generate` should generate tests as well, but we need to standardize/codify our approach to testing before we can do that. We definitely would not use the current qunit/phantom setup that Ember uses since Fastboot code should be tested inside of a Node environment.

Good job. Thanks for looking at this @kratiahuja!

**Comment by kratiahuja** [comment]
@arjansingh : Thanks for your feedback! Here are my responses:

> The new directory structure would be a big breaking change for Ember Apps. Does that mean we're at 2.0.0 already?
I would prefer to hold up 1.0 release in order to pull this in. I can guarantee that this is a blocker for us as well and I would rather introduce this pre 1.0. If there is an issue of resources to work on, we can definitely help. I know @tomdale doesn't want to wait too long but I think we need to push this before 1.0 goes out of the door. As a workaround, we could do symlinks but that is still a bad experience IMO.

> "Explicitly disallow styles in fastboot/"... I'm not sure I understand the issue. Could you elaborate?
CSS is not loaded in fastboot environment. Considering we now have a fastboot namespace, we need to explicitly throw an error from `ember-cli` if someone defines `fastboot/app/styles`. More like a nice to have in the system.

> Our testing strategy for FastBoot has been a little ad-hoc. I do think ember generate should generate tests as well, but we need to standardize/codify our approach to testing before we can do that. We definitely would not use the current qunit/phantom setup that Ember uses since Fastboot code should be tested inside of a Node environment.
I strongly agree with this. I think the long term approach should be to expose Ember test helpers that test your routes in fastboot/node environment. We have been wanting to do this internally for a while since apps bite themselves in production when not having CI with fastboot. However, we haven't had the cycles to do it. Having said that, this (generating tests using `ember generate` should not be a blocker for Fastboot 1.0.

**Comment by kratiahuja** [comment]
@rwjblue Would love your feedback on this issue.

**Comment by tomdale** [comment]
Thank you for the detailed writeup! This definitely aligns with my thinking and am very happy to have you tackle it. Here are a few initial thoughts after reading the above.

#### 2.0?

I think most people on the core team believe that it is worth blocking 1.0 if we can get this in before the Ember 2.10 is out. The reason is simple: currently, just installing the FastBoot addon doubles your build times. That is a pretty awful user experience, and with @kratiahuja's proposed changes, any increase in build time for FastBoot builds should be so negligible most people wouldn't even notice it.

Given that FastBoot is most compatible with Glimmer 2 and its new semantics, and G2 is not in the 2.9 stable release, I think we really have to wait for 2.10 _anyway_; so we may as well take the opportunity to clean up the build process.

That said, it's not 100% clear to me that we cannot implement this plan backwards compatibly. It seems like we could detect if the older folders exist, emit a warning, and create a (possibly expensive) Broccoli tree that filters them _out_ of the primary build but includes them in the secondary, FastBoot-only build. Worst case scenario, we should be able to write a simple migrator that moves things to the right location.

#### Bundle Builds

One thing that we are looking into adopting is a "bundle build"—a build where runtime information about modules is removed and modules are inlined at the correct location so that they are available to the things that depend on them below. See e.g. [the output for this example](http://rollupjs.org/?version=0.36.3&shareable=JTdCJTIyb3B0aW9ucyUyMiUzQSU3QiUyMmZvcm1hdCUyMiUzQSUyMmVzJTIyJTJDJTIybW9kdWxlTmFtZSUyMiUzQSUyMm15QnVuZGxlJTIyJTJDJTIyZ2xvYmFscyUyMiUzQSU3QiU3RCU3RCUyQyUyMm1vZHVsZXMlMjIlM0ElNUIlN0IlMjJuYW1lJTIyJTNBJTIybWFpbi5qcyUyMiUyQyUyMmNvZGUlMjIlM0ElMjJpbXBvcnQlMjAqJTIwYXMlMjBjb25zdGFudHMlMjBmcm9tJTIwJy4lMkZjb25zdGFudHMnJTNCJTVDbiU1Q24lMkYlMkYlMjBJbiUyMHNvbWUlMjBjYXNlcyUyQyUyMHlvdSUyMGRvbid0JTIwa25vdyUyMHdoaWNoJTIwZXhwb3J0cyUyMHdpbGwlNUNuJTJGJTJGJTIwYmUlMjBhY2Nlc3NlZCUyMHVudGlsJTIweW91JTIwYWN0dWFsbHklMjBydW4lMjB0aGUlMjBjb2RlLiUyMEluJTVDbiUyRiUyRiUyMHRoZXNlJTIwY2FzZXMlMkMlMjBSb2xsdXAlMjBjcmVhdGVzJTIwYSUyMG5hbWVzcGFjZSUyMG9iamVjdCU1Q24lMkYlMkYlMjBmb3IlMjBkeW5hbWljJTIwbG9va3VwJTVDbk9iamVjdC5rZXlzKCUyMGNvbnN0YW50cyUyMCkuZm9yRWFjaCglMjBrZXklMjAlM0QlM0UlMjAlN0IlNUNuJTVDdGNvbnNvbGUubG9nKCUyMCU2MFRoZSUyMHZhbHVlJTIwb2YlMjAlMjQlN0JrZXklN0QlMjBpcyUyMCUyNCU3QmNvbnN0YW50cyU1QmtleSU1RCU3RCU2MCUyMCklM0IlNUNuJTdEKSUzQiUyMiU3RCUyQyU3QiUyMm5hbWUlMjIlM0ElMjJjb25zdGFudHMuanMlMjIlMkMlMjJjb2RlJTIyJTNBJTIyZXhwb3J0JTIwY29uc3QlMjAlQ0YlODAlMjAlM0QlMjAzLjE0MTU5JTNCJTVDbmV4cG9ydCUyMGNvbnN0JTIwZSUyMCUzRCUyMDIuNzE4MjglM0IlNUNuZXhwb3J0JTIwY29uc3QlMjAlQ0YlODYlMjAlM0QlMjAxLjYxODAzJTNCJTVDbmV4cG9ydCUyMGNvbnN0JTIwJUNFJUJCJTIwJTNEJTIwMS4zMDM1NyUzQiUyMiU3RCU1RCU3RA==).

Being able to do bundle builds like this is important because it significantly cuts down parse/evaluation time, which is a big bottleneck for larger apps. How do you merge FastBoot modules on top of the primary build if we moved to bundle builds? I'm thinking we can leave some sort of annotation in place until a final pass to remove and minify, but curious to see if you've thought about this more deeply.

#### Browser Only Code

How do you deal with the case where you have some code that accesses e.g. `document` or `window` in the browser and there is no Node-specific version of that file? Put another way, is it possible to say "this module should be included in the browser build, but removed from the FastBoot build?"

#### Separate Port

> Allow existing ember commands to take in additional options. This is needed for fastboot to run on a different port by default.

What is the use case for this? Can't we serve FastBoot requests from the same port as we currently serve `dist/` out of?

####

**Comment by kratiahuja** [comment]
@tomdale: Thanks for you feedback! I really appreciate it! 😄 

> 2.0?

I agree this issue should be fixed ASAP. I promise I am working 100% of my time on it except for dancing here and there once in a while 😃 . I would prefer to wait to cut Fastboot 1.0 until this is merged in. 

I think we can definitely mimick the fastboot command to be doing what we are doing here in `ember-cli` of generating the additive asset and throw a big warning. But eventually the goal should be to kill `ember fastboot` since once this issue is resolved you should be able to do `ember build` and `ember serve` and have a custom fastboot middleware that will be responsible for serving the index.html.  

I will add a corresponding migration task in `ember-cli-fastboot` once this is closer to be merged in. 

> Bundle Builds

I haven't thought much on this since I wasn't aware of the exact timelines from `ember-cli` stand point. I know @nathanhammond has been pushing this at LinkedIn. 

If I understand bundled builds correctly and imagine how they will be integrated in `ember-cli`, they should work for most of fastboot usecases as is when rollup is introduced in `ember-cli` since it will be able to statically get the dependency graph correctly by also walking the fastboot tree.  Though I agree, we should add an annotation in the fastboot doc about this and potentially somewhere in ember-cli docs. 

> Browser Only Code

So it is possible if you manually add the same file in `fastboot/app` and override the behavior of that entire service/anything else. This is because when the assets will be loaded in the sandbox they will be loaded in a particular order: browser assets followed by fastboot assets. So when the assets are eval'd the loader should be able to replace the browser module id with the fastboot specific one (as far as the id will be same). When fastboot-app.js will be generated it will have the module Id as `appName/initializers/ajax` for example. 

Having said that, if the eval of the browser asset fails in the sandbox due to window/document access during eval it will fail fast. If it fails during run time, the [onError](https://github.com/ember-fastboot/ember-cli-fastboot/blob/master/app/initializers/fastboot/error-handler.js) hook will catch it. I believe this is a problem currently in the fastboot builds today as well. I believe if there is a `window/document` access that is any of the hooks that is run during fastboot should be wrapped in by the `isFastboot` check. 

> Separate Port

Sorry this was a typo on my end. I meant to say we (internally at LinkedIn) have a use case where we want to pass it additional command line options as well (for example the port number where are play app should run). I'll update the description with this too. 

Once we are able to expose hooks from `broccoli-middleware` to allow any middleware to serve the base page, we should simply be able to use `ember serve` for fastboot 🎉 .  @tomdale  Just curious to know why does fastboot serve at port 3000? Is there some history behind it?

**Comment by stefanpenner** [comment]
> Separate Port

ya, this seems legacy I dont think we need this going-forward.

**Comment by kellyselden** [comment]
Notes:

* This should be promoted to RFC so we can get line comments.
* This covers work already in progress in other RFCs
  * https://github.com/ember-cli/rfcs/pull/35 covers the generations of “platforms” (fastboot, cordova, electron)
  * https://github.com/ember-cli/rfcs/pull/6 covers extending built in commands
  * https://github.com/ember-cli/rfcs/issues/26 covers build targets (fastboot, cordova, electron)
  * Some of the above have dwindled or been closed, but they should either be revived, or borrowed from into new RFCs
* We should make sure this RFC is compatible with
  * https://github.com/emberjs/rfcs/pull/119 fastboot testing should align with any future testing unification
  * https://github.com/emberjs/rfcs/pull/143 since there’s new folder structure proposed, it should be considered where it fits into the new folder structure incoming.
 * I think a new RFC should exist with any new tree hooks that are needed.
 * I think the ability to modify or replace the dev web server should be an RFC

Basically I see this as getting everything in ember-cli extendable so that `ember-cli-fastboot` can do everything it wants. I don’t think everything in `ember-cli-fastboot` should be ported into ember-cli. This alienates works like `ember-cordova` and `ember-electron` that have wanted this extendibility for a long time.

Thoughts?

**Comment by kratiahuja** [comment]
@kellyselden Thanks for your feedback.

1. Promote to RFC: I know I should have done that earlier, but I don't want to lose the comments now. 
2. Creating a generic platform: I don't want to create the concept of "platforms" right now for two reasons:
- This is a *major* blocker for Fastboot 1.0 and I would like to get this in ASAP. I think getting this in `ember-cli` will be the fastest way.
- I don't know what all hooks other platforms need so I don't want to jump to a generic solution until we have evaluated other platforms and fastboot.
I spoke to Stef about this also and my first pass will be in ember-cli. We can decide from there on if it is generic enough that we can just provide the hooks and move this out of ember-cli. 

Eventually I want to reach to the point you are describing but creating platforms should not be a breaking change for fastboot users so I would like to hold that off and get fastboot building in ember-cli.

**Comment by stefanpenner** [comment]
> I spoke to Stef about this also and my first pass will be in ember-cli. We can decide from there on if it is generic enough that we can just provide the hooks and move this out of ember-cli.

There might be some misunderstanding.

I would like a POC as a PR to ember-cli, we can tease it apart as needed before anything lands in a final release. Regardless of where this work lands, it will need change across many projects, and may ultimately take a different form then the original POC PR.

Lets not get ahead of ourselves, but lets get:

* proof of concept PR that accomplishes a demo with as little friction as possible, this3 lets us concretely figure out what needs to be done.
* move the above proposal to an RFC, so we can iterate as we learn.

**Comment by stefanpenner** [comment]
> Promote to RFC: I know I should have done that earlier, but I don't want to lose the comments now.

* link back to this issue
* incorporate import discussions here into the RFC

**Comment by stefanpenner** [comment]
> This is a major blocker for Fastboot 1.0 and I would like to get this in ASAP.

We wont rush anything until its ready.

> I think getting this in ember-cli will be the fastest way.

That would take up to 12 weeks. I suggested POC to ember-cli, so we can talk about something concrete and then figure out what the next steps are. Those are most likely:

1. changes to ember-cli
2. changes to fastboot
3. a backporting add-on so fastboot can benefit as soon as possible

--- 

Before we get side-tracked with process and bureaucracy lets get a POC out ;)

**Comment by kellyselden** [comment]
> Please note, for some reason RFC process seems to encourage a very waterfall development style. This is not the goal nor the intent, both the POC and RFC should evolve together, as time as demonstrated very few POC or RFC end up landing as is.

I agree, sometimes RFCs stagnate. But I intend to help this issue along the way because I really want it solved as well!

**Comment by kratiahuja** [comment]
@stefanpenner @kellyselden  My apologies for misunderstanding the goal here. 

I am working on cleaning up my POC (in ember-cli) and adding some tests so that it is easier to play with. I will send out a `WIP` PR by end of today. 

I am happy to move this to an RFC but the RFC as the first iteration will only focus on building fastboot needs. I would like to evolve it to something more generic once we know the needs for electron and cordova builds. ETA for an RFC is by end of tomorrow. 

Once I have both the above things upstream, I will close this issue.

**Comment by kratiahuja** [comment]
@stefanpenner @kellyselden : Here is a POC PR: https://github.com/ember-cli/ember-cli/pull/6395 (Apologies if I did something very dumb, this is my first PR in ember-cli and I am still learning).

I will open an RFC tomorrow with more finer details.

**Comment by rwjblue** [comment]
> Apologies if I did something very wrong. This is my first PR in ember-cli and I am still learning

No need to apologize.  I have done many many many PR's to ember-cli and we are still discovering all the dumb things I have done (just ask @nathanhammond sometime). 😝 

Thanks for the WIP PR, will try to review and play with it tonight...

**Comment by kratiahuja** [comment]
RFC as requested is here: https://github.com/ember-cli/rfcs/pull/75 . Going to close this issue.

**Comment by kratiahuja** [comment]
Going to close this issue. The follow action items have been addressed:
1. Short term RFC to allow fastboot to be built inside `ember-cli` is [here](https://github.com/ember-cli/rfcs/pull/77)
2. Corresponding PR that addresses point number 1 is [here](https://github.com/ember-cli/ember-cli/pull/6395)
3. Long term RFC to make this a generic feature of `ember-cli` is [here](https://github.com/ember-cli/rfcs/pull/75)

Point # 1 and # 2 is all that is needed to help unblock release of Fastboot 1.0.
